### DISCUSSION TITLE
Issues with KC25 and argon2

### DISCUSSION THREAD

**Discussion by antikalk** [keycloak/keycloak#32213] [Feedback]
First of all: Thank you for having native support for argon2 finally in Keycloak, this one is very much appreciated!

tl;dr: You will face issues with Keycloak 25 when you already had a custom PasswordHashProvider with the Provider ID **argon2** in use.

In our project we did already have a custom "argon2" provider for quite some time now, allowing for some realms to already use the argon2 password algorithm. Unfortunately this one came with the same provider ID as the build-in Keycloak provider now comes with. This is also the case with this extension here: https://github.com/dreezey/argon2-password-hash-provider

**Problem**:
Provided a user that has already an argon2 credential, being hashed with the custom provider with versions prior to KC 25: The issue with this is, that when you now update to Keycloak 25, the new keycloak password provider is used on login and as some of the required parameters are not provided in the stored credential, the password verification will fail.

There is a workaround: you could just add the order() to your custom Provider to make sure that one is used: but this would still require you to have the custom extension and passwords hashed with the new provider (e.g. if you already deployed KC25 not knowing about this potential issue) will then again not work.

Have other people already experienced this issue and/or have maybe better ideas of how to work with that issue?
